Support booleans and numbers for query values

maksimowiczm opened the pull request:
## Description

Booleans, numeric and nulls are now supported for query values.

#### Changes:
- Implemented ```Deserialize``` manually for ```Template``` using ~~```serde_yaml::value::Value```~~ ```TemplateVisitor```
- Adjusted documentation, although  ```"true"``` is still supported I replaced it with ```true```
- Updated ```CHANGELOG.md``` (twice 😅)

## QA
- Added Template deserialization tests inside ```cereal.rs```

Closes #141

LucasPickering commented on CHANGELOG.md:
```diff
@@ -29,6 +29,7 @@
 - Request recipes can now be organized into folders ([#60](https://github.com/LucasPickering/slumber/issues/60))
   - See [the docs](https://slumber.lucaspickering.me/book/api/request_collection/request_recipe.html#folder-fields) for usage examples
 - Add `slumber show config` and `slumber show collection` subcommands
+- Support booleans and numbers for query values ([#141](https://github.com/LucasPickering/slumber/issues/141))
```
This should be under the Top `Unreleased` section, you have it under 0.18.0. Maybe you added it before the 0.18 release and it stayed here during a rebase? Feel free to add a `Added` section to `Unreleased` for this.

LucasPickering commented on src/collection/cereal.rs:
```diff
@@ -62,6 +66,28 @@ where
+    where
+        D: Deserializer<'de>,
+    {
+        let value: Value = Deserialize::deserialize(deserializer)?;
```
I'd prefer to keep this format-agnostic, in case we want to support multiple config formats in the future. It's possible using `serde_yaml` would still work with other formats, but it's not best practice. I implemented basically the [same logic in another project using a macro](https://github.com/LucasPickering/env-select/blob/master/src/config/cereal.rs#L22), you could basically copy-paste that. Instead of `ValueSource::from_literal` you can just do `value.to_string()` then parse that as a template, like you are now.

Sorry I didn't link this file when I wrote the issue, if I remembered I had that logic already written somewhere I probably would've just done it myself 😛

LucasPickering commented on src/template.rs:
```diff
@@ -58,9 +58,9 @@ pub struct TemplateContext {
 }
 
-/// A immutable string that can contain templated content. The string is parsed
+/// An immutable string that can contain templated content. The string is parsed
```
Thank you!

LucasPickering commented on src/collection/cereal.rs:
```diff
@@ -62,6 +66,28 @@ where
+        Template::try_from(str_value).map_err(|e| Error::custom(e.to_string()))
+    }
+}
+
```
Could you add some tests for this please? I just wrote up a quick one, if you just add some more test cases to cover the other types to it I think that'll be good. If you want you can add a second test to cover error cases as well.

```rust
#[cfg(test)]
mod tests {
    use crate::template::Template;
    use rstest::rstest;
    use serde_test::{assert_de_tokens, Token};

    #[rstest]
    #[case(Token::Str("true"), "true")]
    #[case(Token::Bool(true), "true")]
    fn test_deserialize_template(#[case] token: Token, #[case] expected: &str) {
        assert_de_tokens(&Template::from(expected), &[token]);
    }
}
```

LucasPickering reviewed (changes requested):
Thank you so much for doing this! Looks great, just needs a bit of work I think

maksimowiczm commented:
I adjusted the code accordingly to your suggestions and updated pr description. Force pushed branch because I wanted to split changes between commits.

maksimowiczm commented:
Forgot to run Rustfmt 😅

LucasPickering commented on src/collection/cereal.rs:
```diff
@@ -62,6 +68,64 @@ where
+            visit_primitive!(visit_f64, f64);
+            visit_primitive!(visit_str, &str);
+
+            visit_null!(visit_none);
+            visit_null!(visit_unit);
```
I think supporting `null` here could potentially be confusing, and treating unit as `null` is especially confusing. Leaving an empty value for a key and having it populate as `"null"` is unintuitive. Having `null` populate as `"null"`  is maybe a bit more intuitive but I still don't like it because it ties us to the YAML format.

My preference would be to have `null` not deserialize at all (so it gives an error), and unit deserializes as the empty string.

I'm not sure if this was in your original revision but if so, sorry I missed it

maksimowiczm replied:
My bad, I should have only done booleans and numbers. I think there should be separate issue regarding null value.

LucasPickering commented on src/collection/cereal.rs:
```diff
@@ -62,6 +68,50 @@ where
+                formatter: &mut std::fmt::Formatter,
+            ) -> std::fmt::Result {
+                formatter.write_str(
+                    "Invalid type, must be a string, number or boolean",
```
SORRY one last thing. This results in a clunky error message:

```
invalid type: unit value, expected Invalid type, must be a string, number or boolean
```

It should be just `string, number, or boolean`

LucasPickering commented:
Thanks again! This is very nice to have. And congrats on the very first contribution to Slumber!
